readers/apiserver/watchlist/relist/relist_test.go

```diff
@@ -174,285 +187,352 @@ func TestRelisterList(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
+			var r *Relister
+			var err error
+			
+			// Setup relister
+			if test.setupClientset != nil {
+				clientset := test.setupClientset()
+				r, err = New(clientset)
+				if err != nil {
+					t.Fatalf("New() error = %v", err)
+				}
+			} else if test.setupPagers != nil {
+				r = &Relister{pagers: test.setupPagers()}
 			}
-			continue
-		}
-
-		// Read all entries from channel
-		var receivedEntries []data.Entry
-		var finalErr error
 
-		for response := range ch {
-			if response.Err != nil {
-				finalErr = response.Err
-				break
+			// Setup context
+			ctx := context.Background()
+			var cancel context.CancelFunc
+			if test.contextSetup != nil {
+				ctx, cancel = test.contextSetup()
 			}
-			receivedEntries = append(receivedEntries, response.V)
-		}
 
-		if ec.continueTest(test.name, test.wantStreamErr, finalErr, t) {
-			continue
-		}
+			ch, err := r.List(ctx, test.rt)
+			if ec.continueTest(test.name, test.wantCallErr, err, t) {
+				if !errors.Is(err, exponential.ErrPermanent) {
+					t.Errorf("wanted error was not permanent")
+				}
+				return
+			}
 
-		if len(receivedEntries) != test.wantEntries {
-			t.Errorf("TestRelisterList(%s): got %d entries, want %d", test.name, len(receivedEntries), test.wantEntries)
-		}
-	}
-}
+			// Use custom validation if provided
+			if test.validateBehavior != nil {
+				test.validateBehavior(ctx, cancel, ch, t)
+				return
+			}
 
-func TestRelisterListPagination(t *testing.T) {
-	pods1 := createTestPods(2)
-	entries1 := createTestEntries(pods1)
+			// Standard validation
+			var receivedEntries []data.Entry
+			var finalErr error
 
-	pods2 := createTestPods(1)
-	entries2 := createTestEntries(pods2)
-
-	callCount := 0
-	paginatedLister := func(ctx context.Context, opts metav1.ListOptions) ([]data.Entry, string, error) {
-		callCount++
-		switch callCount {
-		case 1:
-			if opts.Continue != "" {
-				t.Errorf("TestRelisterListPagination: first call should have empty continue token, got %s", opts.Continue)
-			}
-			return entries1, "page2", nil
-		case 2:
-			if opts.Continue != "page2" {
-				t.Errorf("TestRelisterListPagination: second call should have continue token 'page2', got %s", opts.Continue)
+			for response := range ch {
+				if response.Err != nil {
+					finalErr = response.Err
+					break
+				}
+				receivedEntries = append(receivedEntries, response.V)
 			}
-			return entries2, "", nil // Empty continue means no more pages
-		default:
-			t.Errorf("TestRelisterListPagination: unexpected call %d to lister", callCount)
-			return nil, "", errors.New("too many calls")
-		}
-	}
 
-	r := &Relister{
-		pagers: map[types.Retrieve]listPage{
-			types.RTPod: paginatedLister,
-		},
-	}
-
-	ctx := context.Background()
-	ch, err := r.List(ctx, types.RTPod)
-	if err != nil {
-		t.Fatalf("TestRelisterListPagination: List() error = %v", err)
-	}
+			if ec.continueTest(test.name, test.wantStreamErr, finalErr, t) {
```

Comment:
nit: it wasn't clear to me what this meant at first, maybe because 'continue' is being used a lot already (continueToken) and because I think using 'continue' in table tests is a bit specific to how you do it, so renaming might be helpful here.